engine/classes/Elgg/Router.php

```diff
@@ -132,34 +202,142 @@ public function route(Request $request) {
+	 * i.e. /blog/owner/{username?}
+	 *
+	 * Wildcard requirements for common named variables such as 'guid' and 'username'
+	 * will be set automatically.
+	 *
+	 * @param string $name   Unique route name
+	 *                       This name can later be used to generate route URLs
+	 * @param array  $params Route parameters
+	 *                       - path : path of the route
+	 *                       - resource : name of the resource view
+	 *                       - defaults : default values of wildcard segments
+	 *                       - requirements : regex patterns for wildcard segment requirements
+	 *                       - methods : HTTP methods
+	 *
+	 * @return Route
+	 * @throws InvalidParameterException
+	 */
+	public function registerRoute($name, array $params = []) {
+
+		$path = elgg_extract('path', $params);
+		$resource = elgg_extract('resource', $params);
+		$handler = elgg_extract('handler', $params);
+
+		if (!$path || (!$resource && !$handler)) {
+			throw new InvalidParameterException(__METHOD__ . ' requires "path" and "resource" parameters to be set');
+		}
+
+		$defaults = elgg_extract('defaults', $params, []);
+		$requirements = elgg_extract('requirements', $params, []);
+		$methods = elgg_extract('methods', $params, []);
+
+		$patterns = [
+			'guid' => '\d+',
+			'group_guid' => '\d+',
+			'container_guid' => '\d+',
+			'owner_guid' => '\d+',
+			'username' => '[\p{L}\p{Nd}._-]+',
+		];
+
+		$path = trim($path, '/');
+		$segments = explode('/', $path);
+		foreach ($segments as &$segment) {
+			// look for segments that are defined as optional with added ?
+			// e.g. /blog/owner/{username?}
+
+			if (!preg_match('/\{(\w*)(\?)?\}/i', $segment, $matches)) {
+				continue;
+			}
+
+			$wildcard = $matches[1];
+			if (!isset($defaults[$wildcard]) && isset($matches[2])) {
+				$defaults[$wildcard] = ''; // make it optional
+			}
+
+			if (array_key_exists($wildcard, $patterns) && !isset($requirements[$wildcard])) {
+				$requirements[$wildcard] = $patterns[$wildcard];
+			}
+
+			$segment = '{' . $wildcard . '}';
+		}
+
+		$path = '/' . implode('/', $segments);
+
+		$defaults['_resource'] = $resource;
+		$defaults['_handler'] = $handler;
+
+		$route = new Route($path, $defaults, $requirements, [], '', [], $methods);
+
+		$this->routes->add($name, $route);
+
+		return $route;
+	}
+
+	/**
+	 * Unregister a route by its name
+	 *
+	 * @param string $name Name of the route
 	 *
 	 * @return void
 	 */
-	public function unregisterPageHandler($identifier) {
-		unset($this->handlers[$identifier]);
+	public function unregisterRoute($name) {
+		$this->routes->remove($name);
+	}
+
+	/**
+	 * Generate a relative URL for a named route
+	 *
+	 * @param string $name       Route name
+	 * @param array  $parameters Query parameters
+	 *
+	 * @return string
+	 */
+	public function generateUrl($name, array $parameters = []) {
+		try {
+			return $this->generator->generate($name, $parameters);
```

Comment:
Needs 3rd argument `UrlGenerator::RELATIVE_PATH` here. That will generate site-relative paths.

Reply:
Nope, will not work.

Reply:
Unless we want to cut into Symfony's classes I think we want UrlGenerator::ABSOLUTE_URL.